src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -1,50 +1,194 @@
+        } else {
+            lastForThisNode.next = newNode;
+        }
+        size++;
+    }
+
+    private void addElementByIndex(T value, int index) {
+        if (index == 0) {
+            Node<T> newNode = new Node<>(null, value, findByIndex(index));
+            findByIndex(index).prev = newNode;
+            first = newNode;
+            size++;
+        } else {
+            Node<T> after = findByIndex(index);
+            Node<T> before = after.prev;
+            Node<T> newNode = new Node<>(before, value, after);
+            after.prev = newNode;
+            before.next = newNode;
+            size++;
+        }
     }
 
     @Override
     public boolean addAll(List<T> list) {
-        return false;
+        if (list.size() == 0) {
+            return false;
+        }
+        for (T value : list) {
+            add(value);
+        }
+        return true;
     }
 
     @Override
     public T get(int index) {
-        return null;
+        return findByIndex(index).element;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        Node<T> setNode = findByIndex(index);
+        T oldValue = setNode.element;
+        setNode.element = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        Node<T> removeNode = findByIndex(index);
+        return removeHelper(removeNode);
     }
 
     @Override
     public boolean remove(T object) {
-        return false;
+        Node<T> currentNode = first;
+        Node<T> removeNode = null;
+        for (int i = 0; i < size; i++) {
+            if (Objects.equals(currentNode.element, object)) {
+                removeNode = currentNode;
+                break;
+            }
+            currentNode = currentNode.next;
+        }
+
+        if (removeNode == null) {
+            return false;
+        }
+        removeHelper(removeNode);
+        return true;
+    }
+
+    private T removeHelper(Node<T> removeNode) {
+        Node<T> before = removeNode.prev;
+        Node<T> after = removeNode.next;
+        final T removeElement = removeNode.element;
+
+        if (before == null) {
+            first = after;
+        } else {
+            before.next = after;
+            removeNode.prev = null;
+        }
+
+        if (after == null) {
+            last = before;
+        } else {
+            after.prev = before;
+            removeNode.next = null;
+        }
+        removeNode.element = null;
+        size--;
+        return removeElement;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private Node<T> findByIndex(int index) {
+        indexOutOfBoundsExceptionCheck(index);
+        if (index == 0) {
+            return first;
+        } else if (index == size - 1) {
+            return last;
+        }
```

Comment:
do we need this checking for first and last elements?

Reply:
done